File: dsl-api/src/main/java/cbs/dsl/api/context/ConditionalStepBuilder.java
package cbs.dsl.api.context;

import cbs.dsl.api.TransactionDefinition;

import java.util.function.BooleanSupplier;
import java.util.function.Consumer;

// TODO: remove
@Deprecated(forRemoval = true)
public interface ConditionalStepBuilder {

  WhenClause when(BooleanSupplier predicate);

  interface WhenClause {
    ConditionalStepBuilder then(Consumer<ConditionalStepBuilder> block);
  }

  WhenClause orWhen(BooleanSupplier predicate);

  void otherwise(Consumer<ConditionalStepBuilder> block);

  void transaction(TransactionDefinition tx);
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/context/ConditionalStepBuilder.kt
package cbs.dsl.api.context

import cbs.dsl.api.TransactionDefinition

interface ConditionalStepBuilder {
  fun `when`(predicate: () -> Boolean): WhenClause

  interface WhenClause {
    infix fun then(block: ConditionalStepBuilder.() -> Unit): ConditionalStepBuilder
  }

  infix fun orWhen(predicate: () -> Boolean): WhenClause
  infix fun otherwise(block: ConditionalStepBuilder.() -> Unit)
  fun transaction(tx: TransactionDefinition)
}
